pkg/engine/executor/parse_logfmt.go

```diff
@@ -0,0 +1,153 @@
+package executor
+
+import (
+	"sort"
+	"unsafe"
+
+	"github.com/apache/arrow-go/v18/arrow"
+	"github.com/apache/arrow-go/v18/arrow/array"
+	"github.com/apache/arrow-go/v18/arrow/memory"
+
+	"github.com/grafana/loki/v3/pkg/logql/log/logfmt"
+)
+
+// BuildLogfmtColumns builds Arrow columns from logfmt input lines
+// Returns the column headers, the Arrow columns, and any error
+func BuildLogfmtColumns(input *array.String, requestedKeys []string, allocator memory.Allocator) ([]string, []arrow.Array) {
+	columnBuilders := make(map[string]*array.StringBuilder)
+	columnOrder := parseKeys(input, requestedKeys, columnBuilders, allocator)
+
+	// Build final arrays
+	columns := make([]arrow.Array, 0, len(columnOrder))
+	headers := make([]string, 0, len(columnOrder))
+
+	for _, key := range columnOrder {
+		builder := columnBuilders[key]
+		columns = append(columns, builder.NewArray())
+		headers = append(headers, key)
+		builder.Release()
+	}
+
+	return headers, columns
+}
+
+// parseKeys discovers columns dynamically as lines are parsed
+func parseKeys(input *array.String, requestedKeys []string, columnBuilders map[string]*array.StringBuilder, allocator memory.Allocator) []string {
+	columnOrder := []string{}
+	var errorBuilder, errorDetailsBuilder *array.StringBuilder
+	hasErrorColumns := false
+
+	for i := 0; i < input.Len(); i++ {
+		line := input.Value(i)
+		parsed, err := TokenizeLogfmt(line, requestedKeys)
+
+		// Handle error columns
+		if err != nil {
+			// Create error columns on first error
+			if !hasErrorColumns {
+				errorBuilder = array.NewStringBuilder(allocator)
+				errorDetailsBuilder = array.NewStringBuilder(allocator)
+				columnBuilders["__error__"] = errorBuilder
+				columnBuilders["__error_details__"] = errorDetailsBuilder
+				columnOrder = append(columnOrder, "__error__", "__error_details__")
+				hasErrorColumns = true
+
+				// Backfill NULLs for previous rows
+				for j := 0; j < i; j++ {
+					errorBuilder.AppendNull()
+					errorDetailsBuilder.AppendNull()
+				}
```

Comment:
small nit: you could use `AppendNulls(n)` here to avoid the loop. (ditto in other places where you append nulls in a loop)

Reply:
if you feel strongly about it for readability, but each call to `AppendNulls` will create it's own loop, so when multiple columns need nulls appended doing it this way is slightly more performant, though it's probably marginal.

Reply:
> if you feel strongly about it 

nah, just a small nit :) I'm fine with it as-is.